diff --git a/src/Middleware.php b/src/Middleware.php
index 6864a108e03d9c5161a3fc0765619a77ab649348..bf788c06978765a64ecde14c7a12189b853b21d8 100644
--- a/src/Middleware.php
+++ b/src/Middleware.php
@@ -82,7 +82,7 @@ class Middleware implements HttpKernelInterface {
    * {@inheritdoc}
    */
   public function handle(Request $request, $type = HttpKernelInterface::MAIN_REQUEST, $catch = TRUE): Response {
-    if (self::$active && $this->client->isConfigured()) {
+    if (self::$active && $this->client->isConfigured() && php_sapi_name() !== 'cli') {
       $ip = $request->getClientIp();
       if ($this->client->verifyIp($ip) === Constants::REMEDIATION_BAN) {
         $this->logger->info('Blocked @ip, details see <a href="@url">dashboard</a>', [
